server/src/main/java/com/defold/extender/services/DefoldSdkService.java

```diff
@@ -109,26 +96,45 @@ public boolean isLocalSdk(final String sdkVersion) {
         return sdkVersion == null || LOCAL_VERSION.equals(sdkVersion) || System.getenv("DYNAMO_HOME") != null;
     }
 
-    public File getSdk(String hash) throws IOException, URISyntaxException, ExtenderException {
-        return isLocalSdk(hash) ? getLocalSdk() : getRemoteSdk(hash);
+    public DefoldSdk getSdk(String hash) throws ExtenderException {
+        if (isLocalSdk(hash)){
+            return getLocalSdk();
+        }
+        // use ConcurrentHashMap with CompletableFuture here to avoid situation when
+        // several builds needs the same defoldsdk which doesn't exist locally. So one build job starts downloading,
+        // all other jobs wait for download complete and all of then continue running.
+        CompletableFuture<DefoldSdk> operation = operationCache.computeIfAbsent(hash, this::getRemoteSdk);
+        try (DefoldSdk sdk = operation.get()){
+            if (sdk == null) {
+                throw new ExtenderException(String.format("The given sdk does not exist: %s", hash));
+            }
+            if (!sdk.isValid()) {
+                throw new ExtenderException(String.format("The given sdk still does not exist: %s", hash));
+            }
+            evictCache();
+            LOGGER.info("Using Defold SDK version {}", hash);
+            return DefoldSdk.copyOf(sdk);
```

Comment:
Make a copy here because in case when several build jobs requested the same sdk which need to be downloaded they will get one `DefoldSdk` object with ref count == 1. But build jobs have different duration and `DefoldSdk` object should be closed at different time. So here make a copy for every build job.
Also initial `DefoldSdk` calls `close()` several time (because several threads calls `getSdk`). `DefoldSdk.close` implements as idempotent method so first invocation releases reference to sdk, all further - no-op.